### PR TITLE
Add test targets to project when generating from `Package.swift` file.

### PR DESCRIPTION
**Pull request by woin2ee**
Resolves <https://github.com/tuist/tuist/issues/6143>

### Short description 📝

Test targets doesn't have own products so they couldn't be assigned any destinations in destinations resolving [logic](https://github.com/tuist/tuist/blob/1468a058fec48ce59de175cadf7401650289b20d/Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift#L464).
To ensure that a test target has correct destinations, assigned the intersection of destinations of the test target's dependency targets.(by this [PR](https://github.com/tuist/tuist/pull/6251)'s linting rule)

### How to test the changes locally 🧐

Run `tuist generate` for the fixture: `spm_package` and check on test targets.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift [hunk 1]
```diff
@@ -3304,6 +3304,251 @@ final class PackageInfoMapperTests: TuistUnitTestCase {
             )
         )
     }
+
+    func testMap_whenSwiftPackageHavingTestTarget() throws {
```
Comment (fortmarek) [inline Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3308]:
```suggestion
    func testMap_whenSwiftPackageHasTestTarget() throws {
```

nit

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift [hunk 2]
```diff
@@ -3304,6 +3304,251 @@ final class PackageInfoMapperTests: TuistUnitTestCase {
...
+        // Given
+        let basePath = try temporaryPath()
+        let sourcesPath = basePath.appending(try RelativePath(validating: "Package/Sources/Target"))
```
Comment (fortmarek) [inline Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3311]:
```suggestion
        let sourcesPath = basePath.appending(components: "Package", "Sources", "Target")
```

nit: this is imho easier to read and does the same

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift [hunk 3]
```diff
@@ -3304,6 +3304,251 @@ final class PackageInfoMapperTests: TuistUnitTestCase {
...
+        try fileHandler.createFolder(sourcesPath)
+        let testsPath = basePath.appending(try RelativePath(validating: "Package/Tests/TargetTests"))
```
Comment (fortmarek) [inline Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3313]: Same as above

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift [hunk 4]
```diff
@@ -3304,6 +3304,251 @@ final class PackageInfoMapperTests: TuistUnitTestCase {
...
+                            "\(testsPath.pathString)/**",
+                        ])),
+                        dependencies: [.target(name: "Target")]
+                    ),
+                ]
+            )
+        )
+    }
+
+    func testMap_whenSwiftPackageHavingTestTargetWithExplicitProductDestinations() throws {
```
Comment (fortmarek) [inline Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3369]:
```suggestion
    func testMap_whenSwiftPackageHasTestTargetWithExplicitProductDestinations() throws {
```

nit

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift [hunk 5]
```diff
@@ -3304,6 +3304,251 @@ final class PackageInfoMapperTests: TuistUnitTestCase {
...
+                            "\(testsPath.pathString)/**",
+                        ])),
+                        dependencies: [.target(name: "Target")]
+                    ),
+                ]
+            )
+        )
+    }
+
+    func testMap_whenSwiftPackageHavingMultiDependencyTestTargetsWithExplicitProductDestinations() throws {
```
Comment (fortmarek) [inline Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3440]:
```suggestion
    func testMap_whenSwiftPackageHasMultiDependencyTestTargetsWithExplicitProductDestinations() throws {
```

**Review by fortmarek — APPROVED** [review]
A couple of nits, but nothing major. I tested the feature locally and everything works as expected :100:

As a follow-up PR, this [issue](https://github.com/tuist/tuist/issues/5912) is quite similar to what you're doing here but it's not resolved by your changes afaict.

File: Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift (hunk 2 above)
Comment (woin2ee) [reply Tests/TuistLoaderTests/SwiftPackageManager/PackageInfoMapperTests.swift:3311]: I agree.

**Comment by woin2ee** [comment]
I'll look into this [issue](https://github.com/tuist/tuist/issues/5912) as another separate PR from this one.

File: Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift [hunk 6]
```diff
@@ -461,40 +465,20 @@ public final class PackageInfoMapper: PackageInfoMapping {
...
             destinations = Set([.mac])
+        case .test:
+            var testDestinations = Set(XcodeGraph.Destination.allCases)
+            let dependencyNames = target.dependencies.map(\.name)
+            for dependencyName in dependencyNames {
+                let dependencyProducts = targetToProducts[dependencyName] ?? Set()
+                let destinations = unionDestinationsOfProducts(dependencyProducts, in: productDestinations)
+                testDestinations.formIntersection(destinations)
+            }
+            destinations = ProjectDescription.Destinations.from(destinations: testDestinations)
```
Comment (danieleformichelli) [inline Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift:476]: Here we are adding test targets to the tuist projects also for external SPM dependencies, isn't it?
I think we shouldn't do it

**Review by danieleformichelli — CHANGES_REQUESTED** [review]

File: Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift (hunk 6 above)
Comment (fortmarek) [reply Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift:476]: I tried this locally and I didn't get test targets for external dependencies. Can you double check @woin2ee?

**Review by danieleformichelli — APPROVED** [review]

File: Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift (hunk 6 above)
Comment (woin2ee) [reply Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift:476]: I didn't get either. Test targets of external SPM dependencies are ignored at the first of this method and I think it works well. I could have added the same logic here, but I chose to clarify it with comment of first.

**Comment by fortmarek** [comment]
Hey @woin2ee 👋 

This is ready to be merged as soon the pipeline succeeds. I think the test is most likely flaky, can you re-run it? As this is coming from a fork, I can't seem to be able to do that :/

**Comment by woin2ee** [comment]
I didn't know there was anything I could do to re-run. Are there any ways other than a commit?

**Pull request merged**